crates/matrix-sdk/src/oidc/mod.rs

```diff
@@ -434,6 +437,157 @@ impl Oidc {
         LoginWithQrCode::new(&self.client, client_metadata, data)
     }
 
+    /// A higher level wrapper around the configuration and login methods that
+    /// will take some client metadata, register the client if needed and begin
+    /// the login process, returning the authorization data required to show a
+    /// webview for a user to login to their account. Call
+    /// [`Oidc::login_with_oidc_callback`] to finish the process when the
+    /// webview is complete.
+    pub async fn url_for_oidc_login(
+        &self,
+        client_metadata: VerifiedClientMetadata,
+        registrations: OidcRegistrations,
+    ) -> Result<OidcAuthorizationData, OidcError> {
+        let issuer = match self.fetch_authentication_issuer().await {
+            Ok(issuer) => issuer,
+            Err(error) => {
+                if error
+                    .as_client_api_error()
+                    .is_some_and(|err| err.status_code == StatusCode::NOT_FOUND)
+                {
+                    return Err(OidcError::MissingAuthenticationIssuer);
+                } else {
+                    return Err(OidcError::UnknownError(Box::new(error)));
+                }
+            }
+        };
+
+        let redirect_uris =
+            client_metadata.redirect_uris.clone().ok_or(OidcError::MissingRedirectUri)?;
+
+        let redirect_url = redirect_uris.first().ok_or(OidcError::MissingRedirectUri)?;
+
+        self.configure(issuer, client_metadata, registrations).await?;
+
+        let mut data_builder = self.login(redirect_url.clone(), None)?;
+        data_builder = data_builder.prompt(vec![Prompt::Consent]);
+        let data = data_builder.build().await?;
+
+        Ok(data)
+    }
+
+    /// A higher level wrapper around the methods to complete a login after the
+    /// user has logged in through a webview. This method should be used in
+    /// tandem with [`Oidc::url_for_oidc_login`].
+    pub async fn login_with_oidc_callback(
+        &self,
+        authorization_data: &OidcAuthorizationData,
+        callback_url: Url,
+    ) -> Result<()> {
+        let response = AuthorizationResponse::parse_uri(&callback_url)
+            .or(Err(OidcError::InvalidCallbackUrl))?;
+
+        let code = match response {
+            AuthorizationResponse::Success(code) => code,
+            AuthorizationResponse::Error(err) => {
+                if err.error.error == AccessDenied {
+                    // The user cancelled the login in the web view.
+                    return Err(OidcError::CancelledAuthorization.into());
+                }
+                return Err(OidcError::Authorization(err).into());
+            }
+        };
+
+        // Given the InvalidState error already exists, I wonder if this was ever
+        // necessary to manually check?
```

Comment:
Should this be marked with `TODO` or similar?

Reply:
Oh I forgot about that. Well now that there's a test for this it should be pretty easy to answer the question! 👀

Reply:
Turns out it was properly handled *if* you hadn't cancelled a previous request. I've clarified in the comment and prepared to expose an extra method for apps to do the right thing.